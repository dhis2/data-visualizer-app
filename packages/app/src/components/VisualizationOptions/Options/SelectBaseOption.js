import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';

import { sGetUiOptions } from '../../../reducers/ui';
import { acSetUiOptions } from '../../../actions/ui';

export const SelectBaseOption = ({ option, value, onChange }) => (
    <FormControl>
        <InputLabel>{option.label}</InputLabel>
        <Select
            displayEmpty={true}
            onChange={event => onChange(event.target.value)}
            value={value}
        >
            {option.items.map(({ id, label }) => (
                <MenuItem key={id} value={id}>
                    {label}
                </MenuItem>
            ))}
        </Select>
    </FormControl>
);

SelectBaseOption.propTypes = {
    option: PropTypes.object,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    onChange: PropTypes.func,
};

const mapStateToProps = (state, ownProps) => ({
    value: sGetUiOptions(state)[ownProps.option.name],
});

const mapDispatchToProps = (dispatch, ownProps) => ({
    onChange: checked =>
        dispatch(acSetUiOptions({ [ownProps.option.name]: checked })),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(SelectBaseOption);
