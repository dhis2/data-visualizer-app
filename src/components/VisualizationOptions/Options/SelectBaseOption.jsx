import { Checkbox, SingleSelectField, SingleSelectOption } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { connect } from 'react-redux'
import { acSetUiOptions } from '../../../actions/ui.js'
import { sGetUiOption, sGetUiDisabledOption } from '../../../reducers/ui.js'
import {
    tabSectionOption,
    tabSectionOptionToggleable,
} from '../styles/VisualizationOptions.style.js'

export const UnconnectedSelectBaseOption = ({
    option,
    label,
    helpText,
    toggleable,
    value,
    onChange,
    disabled,
    dataTest,
}) => {
    const defaultValue = option.defaultValue
    const [checked, setChecked] = useState(value !== defaultValue)
    const selected = option.items.find((item) => item.value === value)?.value
    const onToggle = (checked) => {
        setChecked(checked)

        onChange(checked ? option.items[0].value : defaultValue)
    }

    return (
        <div className={tabSectionOption.className}>
            {toggleable ? (
                <Checkbox
                    checked={checked}
                    label={label}
                    name={`${option.name}-toggle`}
                    onChange={({ checked }) => onToggle(checked)}
                    dense
                    disabled={disabled}
                    dataTest={`${dataTest}-checkbox`}
                />
            ) : null}
            {!toggleable || checked ? (
                <div
                    className={
                        toggleable ? tabSectionOptionToggleable.className : ''
                    }
                >
                    <SingleSelectField
                        name={`${option.name}-select`}
                        label={toggleable ? '' : label}
                        onChange={({ selected }) => onChange(selected)}
                        selected={selected}
                        helpText={helpText}
                        inputWidth="280px"
                        dense
                        dataTest={`${dataTest}-select`}
                        disabled={disabled}
                    >
                        {option.items.map(({ value, label }) => (
                            <SingleSelectOption
                                key={value}
                                value={value}
                                label={label}
                                dataTest={`${dataTest}-option`}
                            />
                        ))}
                    </SingleSelectField>
                </div>
            ) : null}
        </div>
    )
}

UnconnectedSelectBaseOption.propTypes = {
    option: PropTypes.object.isRequired,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    onChange: PropTypes.func.isRequired,
    dataTest: PropTypes.string,
    disabled: PropTypes.bool,
    helpText: PropTypes.string,
    label: PropTypes.string,
    toggleable: PropTypes.bool,
}

const mapStateToProps = (state, ownProps) => ({
    disabled: Boolean(sGetUiDisabledOption(state, ownProps.option)),
    helpText:
        sGetUiDisabledOption(state, ownProps.option)?.helpText ||
        ownProps.helpText,
    value: sGetUiOption(state, ownProps.option),
})

const mapDispatchToProps = (dispatch, ownProps) => ({
    onChange: (value) =>
        dispatch(acSetUiOptions({ [ownProps.option.name]: value })),
})

export const SelectBaseOption = connect(
    mapStateToProps,
    mapDispatchToProps
)(UnconnectedSelectBaseOption)
