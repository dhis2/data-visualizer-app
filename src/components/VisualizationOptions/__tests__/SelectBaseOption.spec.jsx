import { render, screen, fireEvent } from '@testing-library/react'
import React from 'react'
import { UnconnectedSelectBaseOption as SelectBaseOption } from '../Options/SelectBaseOption.jsx'

jest.mock('@dhis2/ui', () => {
    const originalModule = jest.requireActual('@dhis2/ui')
    return {
        ...originalModule,
        SingleSelectOption: ({ value, label }) => (
            <option value={value} data-test={`option-${value}`}>
                {label}
            </option>
        ),
        SingleSelectField: ({ children, dataTest, label, name, onChange }) => (
            <select
                label={label}
                name={name}
                data-test={dataTest}
                onChange={onChange}
            >
                {children}
            </select>
        ),
    }
})

describe('DV > Options > SelectBaseOption', () => {
    const onChange = jest.fn()

    const props = {
        value: '',
        label: 'toggleable test',
        option: {
            name: 'toggleable',
            defaultValue: '',
            items: [
                { value: 'opt1', label: 'Option 1' },
                { value: 'opt2', label: 'Option 2' },
                { value: 'opt3', label: 'Option 3' },
            ],
        },
        onChange,
        toggleable: true,
        dataTest: 'hello',
    }

    test('renders a <Checkbox />', () => {
        const { queryByTestId } = render(<SelectBaseOption {...props} />)
        const checkbox = screen.getByRole('checkbox', {
            name: props.label,
        })
        expect(checkbox).toBeInTheDocument()
        expect(checkbox).toHaveProperty('checked', false)

        const select = queryByTestId('hello-select')
        expect(select).not.toBeInTheDocument()
    })

    test('renders a <Checkbox /> and <SingleSelectField /> when the checkbox is enabled', () => {
        props.value = 'opt1'

        const { container, queryByTestId } = render(
            <SelectBaseOption {...props} />
        )
        expect(container).toMatchSnapshot()

        const checkbox = screen.getByRole('checkbox', {
            name: props.label,
        })
        expect(checkbox).toBeInTheDocument()

        const select = queryByTestId('hello-select')
        expect(select).toBeInTheDocument()
    })

    test('renders a <SingleSelectField /> when not toggleable', () => {
        props.toggleable = false
        const { container, queryByTestId } = render(
            <SelectBaseOption {...props} />
        )
        expect(container).toMatchSnapshot()

        const select = queryByTestId('hello-select')
        expect(select).toBeInTheDocument()

        const options = Array.from(select.children)
        expect(options).toHaveLength(3)
        options.forEach((option, index) => {
            expect(option.getAttribute('value')).toEqual(`opt${index + 1}`)
        })
    })

    test('should trigger the onChange callback on select change', () => {
        props.toggleable = false
        const { queryByTestId } = render(<SelectBaseOption {...props} />)

        const select = queryByTestId('hello-select')
        expect(select).toBeInTheDocument()

        fireEvent.change(select, { target: { value: 'opt2' } })

        expect(onChange).toHaveBeenCalled()
        // expect(onChange).toHaveBeenCalledWith('opt2')
    })
})
