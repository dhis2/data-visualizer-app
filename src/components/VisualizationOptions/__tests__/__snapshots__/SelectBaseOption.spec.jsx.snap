// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DV > Options > SelectBaseOption renders a <Checkbox /> and <SingleSelectField /> when the checkbox is enabled 1`] = `
<div>
  <div>
    <label
      class="jsx-1193297444 dense"
      data-test="hello-checkbox"
    >
      <input
        checked=""
        class="jsx-1193297444"
        name="toggleable-toggle"
        type="checkbox"
        value=""
      />
      <div
        class="jsx-1193297444 icon dense"
      >
        <svg
          class="jsx-4069423049 jsx-3997907426 checked"
          viewBox="0 0 14 14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            class="jsx-4069423049 jsx-3997907426 background"
            height="14"
            rx="2"
            width="14"
            x="0"
            y="0"
          />
          <path
            class="jsx-4069423049 jsx-3997907426 border"
            d="M2,0 L12,0 C13.1045695,0 14,0.8954305 14,2 L14,12 C14,13.1045695 13.1045695,14 12,14 L2,14 C0.8954305,14 0,13.1045695 0,12 L0,2 C0,0.8954305 0.8954305,0 2,0 Z M2,1 C1.44771525,1 1,1.44771525 1,2 L1,12 C1,12.5522847 1.44771525,13 2,13 L12,13 C12.5522847,13 13,12.5522847 13,12 L13,2 C13,1.44771525 12.5522847,1 12,1 L2,1 Z"
          />
          <path
            class="jsx-4069423049 jsx-3997907426 checkmark"
            d="M10.3520005,3.30015877 L5.54268293,8.03252726 L3.64634146,6.16652726 L2.28668615,7.5044281 L5.54310367,10.7010587 L11.7145993,4.628307 L10.3520005,3.30015877 Z M5.54268293,9.43547274 L5.61193184,9.36733181 L5.54226219,9.29894127 L5.47309681,9.367 L5.54268293,9.43547274 Z M10.6496329,4.41023392 L10.6495057,4.35652297 L10.5648412,4.35672356 L10.2854007,4.631693 L10.3553166,4.69984123 L10.6496329,4.41023392 Z M3.42891348,7.22439304 L3.35030905,7.22439304 L3.35034135,7.27820863 L3.64634146,7.56947274 L3.71331385,7.5035719 L3.42891348,7.22439304 Z"
          />
          <rect
            class="jsx-4069423049 jsx-3997907426 indeterminate"
            height="2"
            width="8"
            x="3"
            y="6"
          />
        </svg>
      </div>
      toggleable test
    </label>
    <div
      class="undefined"
    >
      <select
        data-test="hello-select"
        label=""
        name="toggleable-select"
      >
        <option
          data-test="option-opt1"
          value="opt1"
        >
          Option 1
        </option>
        <option
          data-test="option-opt2"
          value="opt2"
        >
          Option 2
        </option>
        <option
          data-test="option-opt3"
          value="opt3"
        >
          Option 3
        </option>
      </select>
    </div>
  </div>
</div>
`;

exports[`DV > Options > SelectBaseOption renders a <SingleSelectField /> when not toggleable 1`] = `
<div>
  <div>
    <div
      class=""
    >
      <select
        data-test="hello-select"
        label="toggleable test"
        name="toggleable-select"
      >
        <option
          data-test="option-opt1"
          value="opt1"
        >
          Option 1
        </option>
        <option
          data-test="option-opt2"
          value="opt2"
        >
          Option 2
        </option>
        <option
          data-test="option-opt3"
          value="opt3"
        >
          Option 3
        </option>
      </select>
    </div>
  </div>
</div>
`;
